src/QMCWaveFunctions/TrialWaveFunction.h

```diff
@@ -282,9 +283,9 @@ class TrialWaveFunction: public MPIObjectBase
   void rejectMove(int iat);
   void acceptMove(ParticleSet& P, int iat);
 
-  RealType registerData(ParticleSet& P, BufferType& buf);
-  RealType updateBuffer(ParticleSet& P, BufferType& buf, bool fromscratch=false);
-  void copyFromBuffer(ParticleSet& P, BufferType& buf);
+  void registerData(ParticleSet& P, WFBufferType& buf);
```

Comment:
It would be good to add some description of these buffer methods - what they do and when should they be called.
My current understanding:
registerData allocates space in the buffer 'buf' based on the data type sizes in the ParticleSet objects.   It should be called before the buffer is used in updateBuffer/copyFromBuffer (?)

Is the buffer only used to for sending walkers across processors?  Or is it also the temporary storage for cached values?

updateBuffer - copies data from the TrialWaveFunction tree to the buffer (?).  The code also appears to update some values in ParticleSet (P.G and P.L), but maybe those values are scratch space used to put the gradient and laplacian in the buffer?

copyFromBuffer - copy data from 'buf' to the tree describing the wavefunction rooted in the TrialWaveFunction object.

Reply:
The basic description is in OrbitalBase class.
registerData requests space in the buffer 'buf' based on the data type sizes of the objects in this class.
After all the registration in walker initialization, buf.allocate() is called to actually allocate the space.
copyFromBuffer copy data or attach memory from 'buf' to the class and updateBuffer put the data or forward the memory cursor.

The buffer is used not only for MPI but more for the case when the number of walkers different from the number of threads.

ParticleSet also registerData, copyFromBuffer, updateBuffer. This is purely for MPI. P.G and P.L are not scratch space but used everywhere in the code.

I agree with you that it will be healthier to have more in-source documentation. I will update them.

Reply:
These comments will be a good addition the documentation for these functions - the context of how a function is used is sometimes the hardest to capture, but also the most valuable for someone trying to understand the code.

It would also be good to add a comment here pointing to OrbitalBase as the location of the documentation.  TrialWaveFunction doesn't inherit from OrbitalBase, so it's not obvious where to find it.